technical/candles.py

```diff
@@ -4,16 +4,30 @@
 
 
 def heikinashi(bars):
+"""
+    Heikin Ashi calculation: https://school.stockcharts.com/doku.php?id=chart_analysis:heikin_ashi
+
+    ha_open calculation based on: https://stackoverflow.com/a/55110393
+    ha_open = [ calculate first record ][ append remaining records with list comprehension method ]
+    list comprehension method is significantly faster as a for loop
+
+    result:
+    ha_open[0] =  (bars.open[0] + bars.close[0]) / 2
+    ha_open[1] = (ha_open[0] + ha_close[0]) / 2
+    ...
+    ha_open[last] = ha_open[len(bars)-1] + ha_close[len(bars)-1]) / 2
+    """
+
     bars = bars.copy()
-    bars['ha_close'] = (bars['open'] + bars['high'] +
-                        bars['low'] + bars['close']) / 4
 
-    bars.at[0, 'ha_open'] = (bars.at[0, 'open'] + bars.at[0, 'close']) / 2
-    for x in range(1, len(bars)):
-        bars.at[x, 'ha_open'] = (bars.at[x - 1, 'ha_open'] + bars.at[x - 1, 'ha_close']) / 2
+    bars['ha_close'] = bars[['open', 'high', 'low', 'close']].mean(axis=1)
+
+    ha_open = [ (bars.open[0] + bars.close[0]) / 2 ]
+    [ ha_open.append((ha_open[x] + bars.ha_close[x]) / 2) for x in range(0, len(bars)-1) ]
+    bars['ha_open'] = ha_open
 
-    bars['ha_high'] = bars.loc[:, ['high', 'ha_open', 'ha_close']].max(axis=1)
-    bars['ha_low'] = bars.loc[:, ['low', 'ha_open', 'ha_close']].min(axis=1)
+    bars['ha_high'] = bars[['ha_open', 'ha_close', 'high']].max(axis=1)
+    bars['ha_low'] = bars[['ha_open', 'ha_close', 'low']].min(axis=1)
```

Comment:
i think we should still use `.loc[]` here (actually it should be used on the left hand side as well).

it's pandas recommended way - and it might be slightly faster as not using .loc has pandas first figure out which way our indexing is intended.

Reply:
okay, thanks for the update. Reviewing the python documentation i see that the colon is used to select all rows, so then I think it should be like this:
````python
    bars.loc[:, 'ha_close'] = bars.loc[:, ['open', 'high', 'low', 'close']].mean(axis=1)

    ha_open = [ (bars.open[0] + bars.close[0]) / 2 ]
    [ ha_open.append((ha_open[x] + bars.ha_close[x]) / 2) for x in range(0, len(bars)-1) ]
    bars['ha_open'] = ha_open

    bars.loc[:, 'ha_high'] = bars.loc[:, ['ha_open', 'ha_close', 'high']].max(axis=1)
    bars.loc[:, 'ha_low'] = bars.loc[:, ['ha_open', 'ha_close', 'low']].min(axis=1)
````
Is that matching with your expectations?

I just tested this, and speed does not make much difference.  But if its the recommended way then we should definitely use it.
````
Without .loc
Pair: BTC-USDT
Technical time = 0.0803 [s]
QtpyLib time   = 0.1186 [s]
Pair: ETH-USDT
Technical time = 0.0746 [s]
QtpyLib time   = 0.1248 [s]
Pair: LTC-USDT
Technical time = 0.0735 [s]
QtpyLib time   = 0.1191 [s]

With .loc
Pair: BTC-USDT
Technical time = 0.0766 [s]
QtpyLib time   = 0.1137 [s]
Pair: ETH-USDT
Technical time = 0.0896 [s]
QtpyLib time   = 0.1158 [s]
Pair: LTC-USDT
Technical time = 0.0739 [s]
QtpyLib time   = 0.1125 [s]
````

Reply:
yes exactly

it'll for sure not make a TON of difference - but it's the "clear" way to do this (.loc[] has rows as first argument, and columns as 2nd argument)

Reply:
I have just added the changes.

Reply:
i added one more small modification. Changing the sequence of `'high', 'ha_open', 'ha_close'` as they exist in the dataframe.
````python
    bars.loc[:, 'ha_high'] = bars.loc[:, ['high', 'ha_open', 'ha_close']].max(axis=1)
    bars.loc[:, 'ha_low'] = bars.loc[:, ['low', 'ha_open', 'ha_close']].min(axis=1)
````